helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ResourceTopStateAntiAffinityConstraint.java

```diff
@@ -29,25 +29,14 @@
  * score, vice versa.
  */
 class ResourceTopStateAntiAffinityConstraint extends SoftConstraint {
-  private static final float MAX_SCORE = 1f;
-  private static final float MIN_SCORE = 0f;
-
-  ResourceTopStateAntiAffinityConstraint() {
-    super(MAX_SCORE, MIN_SCORE);
-  }
 
   @Override
   protected float getAssignmentScore(AssignableNode node, AssignableReplica replica,
       ClusterContext clusterContext) {
-    if (!replica.isReplicaTopState()) {
```

Comment:
This means if the replica is not with the top state, this constraint should not score differently. Or the node with more master will tend to have less replicas (any state) because of this constraint.

Please keep the main logic unchanged here. Just update how we calculate the percentage.

Reply:
Does it mean if a replica is not in top state, as long as the constraint returns the same value (the value could just be anything, e.g, 0), it will not affect the algorithm result. Only evaluate when the replica is top state, since it makes the “top state” situations worse for a node.

Reply:
The point is the constraint should return the same value across all the nodes if the input replica is not the top state. Meaning all 0, or some random numbers. However, your change will return the value depends on getAssignedTopStatePartitionsCount(), this value is different from node to node.

Reply:
Return 0 as the default score to make it easier and added some comments on top of it. Merged with the latest changes.